unitary/alpha/quantum_world.py

```diff
@@ -106,29 +127,106 @@ def combine_with(self, other_world: "QuantumWorld"):
+
+        Args:
+            namespace: Custom string to be added in the name
+            value: The value for the ancilla qudit
+
+        Returns:
+            The added ancilla object.
+        """
+        count = 0
+        ancilla_name = f"ancilla_{namespace}_{count}"
+        while ancilla_name in self.object_name_dict:
+            count += 1
+            ancilla_name = f"ancilla_{namespace}_{count}"
+        new_obj = QuantumObject(ancilla_name, value)
+        self.add_object(new_obj)
+        self.ancilla_names.add(ancilla_name)
+        return new_obj
+
     def _append_op(self, op: cirq.Operation):
         """Add the operation in a way designed to speed execution.
 
         For the sparse simulator post-selections should be as early as possible to cut
         down the state size. Also X's since they don't increase the size.
         """
-        if not self.use_sparse:
-            self.circuit.append(op)
-            return
 
-        if isinstance(op, PostSelectOperation) or op.gate is cirq.X:
+        if not self.use_sparse or isinstance(
+                op, PostSelectOperation) or op.gate is cirq.X:
             strategy = cirq.InsertStrategy.EARLIEST
         else:
             strategy = cirq.InsertStrategy.NEW
+
+        if self.compile_to_qubits:
+            op = self._compile_op(op)
+
         self.circuit.append(op, strategy=strategy)
 
+    def _get_num_compiled_qubits(self, qudit: cirq.Qid,
+                                 num_expected_qubits: int) -> List[cirq.Qid]:
+        """Gets the exact number of compiled qubits for the qudit.
+
+        May create extra ancilla qubits, if needed. Currently, none of the
+        non-trivial cases are exercised because our actual qubit compilation
+        does not support mixing of different qudit dimensions.
+        """
+        num_compiled_qubits = len(self.compiled_qubits[qudit])
+        if num_expected_qubits > num_compiled_qubits:
+            # Original padding was not enough. This means that one
+            # of the qubits in the Operation has a higher dimension
+            # than this qubit.
+            for qubit_num in range(num_compiled_qubits, num_expected_qubits):
+                new_obj = self._add_ancilla(qudit.name)
+                self.compiled_qubits[qudit].append(new_obj.qubit)
+            return self.compiled_qubits[qudit]
+        if num_expected_qubits < num_compiled_qubits:
+            # This op needs less padding than what this qubit has.
+            # Use the first N qubits based on the size required.
+            return self.compiled_qubits[qudit][:num_expected_qubits]
+        else:
+            return self.compiled_qubits[qudit]
+
+    def _compile_op(self, op: cirq.Operation) -> cirq.Operation:
+        """Compiles the operation down to qubits, if needed."""
+        qid_shape = cirq.qid_shape(op)
+        distinct_dims = set(qid_shape)
+        if len(distinct_dims) > 1:
+            # TODO: Add support for arbitrary Qid shapes to
```

Comment:
Consider linking this to a github issue. (or just remove the comment)

Reply:
Done